python_scripts/load_arrow.py

```diff
@@ -1,13 +1,295 @@
+        Columns are immutable, so first find the index of the original column,
+        remove it, and replace it with the new column at the same index and with the
+        same name.
+        """
+        col_name_index = self.column_names.index(column_name)
+        self.arrow_table = self.arrow_table.remove_column(col_name_index)
+        self.arrow_table = self.arrow_table.add_column(
+            col_name_index, column_name, converted
+        )
+
+    def replace_missing_values(self):
+        """
+        Replace null values in the arrow table with the value stata needs.
+        Dates/timestamps/category require additional conversion before loading to
+        stata, so null replacements are deferred till later.
+        """
+        for column_name, column_type in self.column_types.items():
+            if column_type in ["date", "timestamp", "category"]:
+                # Skip dates/timestamps/category
+                # We replace these after conversion to stata dates in `load_data`
+                continue
+            null_value = self.MISSING_VALUES[column_type]
+            converted = self.arrow_table[column_name].fill_null(null_value)
+            col_name_index = self.column_names.index(column_name)
+            self.arrow_table = self.arrow_table.remove_column(col_name_index)
+            self.arrow_table = self.arrow_table.add_column(
+                col_name_index, column_name, converted
+            )
+
+    def get_categorical_encodings(self, varnames):
+        """
+        Retrieve a mapping of labels: values for categorical vars.
+        Returns a dict where the variable names are the keys and the value
+        is a dictionary consisting of label : value key/value pairs
+        """
+        value_labels = {}
+        for varname in varnames:
+            first_chunk = self.arrow_table[varname].chunks[0]
+            # Get the unique values for this variable and numeric mapping
+            value_map = {
+                str(value): idx for idx, value in enumerate(first_chunk.dictionary)
+            }
+            value_labels[varname] = value_map
+
+        return value_labels
+
+    def make_vars(self):
+        """
+        Function used to create the Stata variables in memory.
+        This assumes that the values in name_and_type object are valid
+        Stata variable names.
+        :param name_and_type: A dictionary structured to contain the variable names
+        as the keys and the storage type of the variable as the value.
+        """
+        # Loop over the variable name / variable type mappings and add
+        # variables with the appropriate typing based on the vartype
+        for varname, vartype in self.column_types.items():
+            if vartype == "string":
+                sfi.Data.addVarStrL(varname)
+            elif vartype == "boolean":
+                sfi.Data.addVarByte(varname)
+            elif vartype == "int":
+                sfi.Data.addVarInt(varname)
```

Comment:
@evansd I am looking at this now, and it is going OK, except for categoricals with missing values.  It's pretty straightforward to find the min/max and then the appropriate Stata type for those.  And for categoricals, we find the size of the dictionary and use that and 0 for min/max, since the values are indices.

But, when a categorical has null values, and we replace that null value in the indices list with the appropriate stata null value (101 for bytes, which is what most categoricals will be), it gets converted to `int` when it's loaded into stata (even though the variable was created as a byte-type var).  Similarly, if I make all categoricals ints, and replace nulls with the int missing value (32768), it gets converted to `long` when it's loaded into stata. Only when we have float/double and the missing value is the stata sysmiss do they seem to get properly encoded as missing when they get into stata.

Any ideas?

Reply:
This is all a bit depressing, isn't it?

The only thing I can think of it is to try to create a categorical variable with missing values from within Stata and then try to reverse engineer what it's done. Possibly by writing it out as a `.dta` file and then loading it into Pandas? Or – worst case – by hex-spelunking the `.dta` file itself to work out how it's been encoded.

Or maybe it will turn out that Stata doesn't really handle missing values in categoricals. We might need to invoke Alex as a our resident Stata person here.

Reply:
Actually I think I was wrong - it's not just categoricals.  If there's an int, and I create a Int-type stata variable for it, it automatically loads it to stata as `long` and treats the null-replacement value as ints.  One thought I had was to load it, and then replace the null-replacement values in stata based on the types we originally define (pre-null replacement)

Reply:
Ah, right. That almost makes more sense, not that it helps us much!

I'm aware that I've been chucking out grand plans in the abstract based on how I _think_ it ought to work without knowing how it actually does work.

Does that mean there's just no way of passing missing values up through the Python interface? That's a bit unexpected, but I suppose it wouldn't be entirely surprising. What happens with Python `None` values?

Reply:
Python None values (if you don't replace them with something) make it error when you try to load it to stata

Reply:
I don't fully understand what this documentation is saying, but is it possible that we could do something with the `selectvar` argument here? Will setting it to `-1` somehow possibly do what we want?
https://www.stata.com/python/api16/Data.html#sfi.Data.store

Reply:
I already tried that, and I don't think so. I couldn't get it to work, I got this error, no matter what I tried.
```
Traceback (most recent call last):
  File "/python_scripts/load_arrow.py", line 455, in <module>
    main(**parse_args())
  File "/python_scripts/load_arrow.py", line 23, in main
    converter.load_data()
  File "/python_scripts/load_arrow.py", line 432, in load_data
    sfi.Data.store(
  File "/usr/local/stata/ado/base/py/sfi.py", line 1431, in store
    return _stp._st_storedata(vars, obss, val, svar)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
SystemError: conformability error; failed to store values in the current Stata 
```
(That's the same error I get if I leave the None values)

I also tried with just one column, in case the -1 option required named variables in `var`, but same error.

I have a solution now, which is to load it into stata and then run stata commands to replace the numeric null value with `.` and cast the variable to what we determined it should be in the first place.